File: leetcode/701-insert-into-a-binary-search-tree/main.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// iterative search
func insertIntoBST(root *TreeNode, val int) *TreeNode {
	if root == nil {
		return &TreeNode{val, nil, nil}
	}
	var curr *TreeNode
	curr = root
	for true {
		if val > curr.Val {
			if curr.Right != nil {
				curr = curr.Right
			} else {
				curr.Right = &TreeNode{val, nil, nil}
				break
			}
		} else {
			if curr.Left != nil {
				curr = curr.Left
			} else {
				curr.Left = &TreeNode{val, nil, nil}
				break
			}
		}
	}
	return root
}

// recursive search
func insertIntoBST1(root *TreeNode, val int) *TreeNode {
	if root == nil {
		return &TreeNode{val, nil, nil}
	}
	var recursion func(curr *TreeNode)
	recursion = func(curr *TreeNode) {
		if val > curr.Val {
			if curr.Right != nil {
				recursion(curr.Right)
			} else {
				curr.Right = &TreeNode{val, nil, nil}
			}
		} else {
			if curr.Left != nil {
				recursion(curr.Left)
			} else {
				curr.Left = &TreeNode{val, nil, nil}
			}
		}
	}
	recursion(root)
	return root
}

// another way:
// unfold the tree into an array,
// construct an binary tree based on the array
// PS1 the tree might be resulted in a different structure
// PS2 too tedious wont implement

func main() {
	// 		4
	//	2		7
	// 1 3
	root := &TreeNode{4,
		&TreeNode{2,
			&TreeNode{1, nil, nil},
			&TreeNode{3, nil, nil},
		},
		&TreeNode{7, nil, nil},
	}
	ans := insertIntoBST1(root, 5)
	println(ans.Right.Left.Val)
}

File: leetcode/701-insert-into-a-binary-search-tree/main.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def insertIntoBST(self, root, val):
        """
        :type root: TreeNode
        :type val: int
        :rtype: TreeNode

        1st approach: iterative search the direction we are going to go down and insert the target into the right position

        104 ms, faster than 63.72%
        """
        if root == None:
            return TreeNode(val)
        cur = root
        while cur:
            if val < cur.val:
                if cur.left == None:
                    cur.left = TreeNode(val)
                    break
                cur = cur.left
            elif val > cur.val:
                if cur.right == None:
                    cur.right = TreeNode(val)
                    break
                cur = cur.right
            else:
                break
        return root


class Solution(object):
    def insertIntoBST(self, root, val):
        """
        :type root: TreeNode
        :type val: int
        :rtype: TreeNode

        2nd approach: recursive search the direction we are going to go down and insert the target into the right position

        104 ms, faster than 71.97%
        """
        if root == None:
            return TreeNode(val)
        if val < root.val:
            root.left = self.insertIntoBST(root.left, val)
        else:
            root.right = self.insertIntoBST(root.right, val)
        return root
